src/common/hist_util.cu

```diff
@@ -98,379 +43,413 @@ __global__ void UnpackFeaturesK(float* __restrict__ fvalues,
+  auto d_temp_weights = temp_weights.data().get();
+  page.offset.SetDevice(device);
+  auto row_ptrs = page.offset.ConstDeviceSpan();
+  size_t base_rowid = page.base_rowid;
+  dh::LaunchN(device, temp_weights.size(), [=] __device__(size_t idx) {
+    size_t element_idx = idx + begin;
+    size_t ridx = thrust::upper_bound(thrust::seq, row_ptrs.begin(),
+                                      row_ptrs.end(), element_idx) -
+                  row_ptrs.begin() - 1;
+    d_temp_weights[idx] = weights[ridx + base_rowid];
+  });
+
+  // Sort
+  thrust::sort_by_key(thrust::cuda::par(alloc), sorted_entries.begin(),
+                      sorted_entries.end(), temp_weights.begin(),
+                      EntryCompareOp());
+  std::vector<Entry> entries_t(sorted_entries.begin(), sorted_entries.end());
+
+  // Scan weights
+  thrust::inclusive_scan_by_key(thrust::cuda::par(alloc),
+                                sorted_entries.begin(), sorted_entries.end(),
+                                temp_weights.begin(), temp_weights.begin(),
+                                [=] __device__(const Entry& a, const Entry& b) {
+                                  return a.index == b.index;
+                                });
+  std::vector<float> weights_t(temp_weights.begin(), temp_weights.end());
+
+  dh::caching_device_vector<size_t> column_sizes_scan;
+  GetColumnSizesScan(device, &column_sizes_scan,
+                     {sorted_entries.data().get(), sorted_entries.size()},
+                     num_columns);
+  thrust::host_vector<size_t> host_column_sizes_scan(column_sizes_scan);
+
+  // Extract cuts
+  dh::caching_device_vector<SketchEntry> cuts(num_columns * num_cuts);
+  ExtractWeightedCuts(
+      device, {cuts.data().get(), cuts.size()}, num_cuts,
+      {sorted_entries.data().get(), sorted_entries.size()},
+      {temp_weights.data().get(), temp_weights.size()},
+      {column_sizes_scan.data().get(), column_sizes_scan.size()});
+
+  // add cuts into sketches
+  thrust::host_vector<SketchEntry> host_cuts(cuts);
+  sketch_container->Push(num_cuts, host_cuts, host_column_sizes_scan);
+}
+
+HistogramCuts DeviceSketch(int device, DMatrix* dmat, int max_bins,
+                           size_t sketch_batch_num_elements) {
+  HistogramCuts cuts;
+  DenseCuts dense_cuts(&cuts);
+  SketchContainer sketch_container(max_bins, dmat->Info().num_col_,
+                                   dmat->Info().num_row_);
+
+  constexpr int kFactor = 8;
+  double eps = 1.0 / (kFactor * max_bins);
+  size_t dummy_nlevel;
+  size_t num_cuts;
+  WQuantileSketch<bst_float, bst_float>::LimitSizeLevel(
+      dmat->Info().num_row_, eps, &dummy_nlevel, &num_cuts);
+  num_cuts = std::min(num_cuts, dmat->Info().num_row_);
+  if (sketch_batch_num_elements == 0) {
+    sketch_batch_num_elements = dmat->Info().num_nonzero_;
+  }
+  dmat->Info().weights_.SetDevice(device);
+  for (const auto& batch : dmat->GetBatches<SparsePage>()) {
+    size_t batch_nnz = batch.data.ConstHostVector().size();
```

Comment:
If the input is from device this will pull all data back to host.  Do you think we can just call `batch.data.Size()`?